src/main/java/com/epam/aidial/core/upstream/UpstreamRoute.java

```diff
@@ -1,76 +1,124 @@
 package com.epam.aidial.core.upstream;
 
 import com.epam.aidial.core.config.Upstream;
-import lombok.RequiredArgsConstructor;
+import com.epam.aidial.core.util.HttpStatus;
+import io.vertx.core.http.HttpClientResponse;
+import lombok.extern.slf4j.Slf4j;
 
-import java.util.Iterator;
-import java.util.List;
-import java.util.Objects;
+import javax.annotation.Nullable;
 
-@RequiredArgsConstructor
-public class UpstreamRoute implements Iterator<Upstream> {
+/**
+ * Not thread-safe! Upstream route with retry management.<br>
+ * This class should be used to route only one user request with retry ability.<br>
+ * Typical usage:
+ * <pre>
+ * if (!available()) {
+ *     return "No route"; // if no upstream available or max retry attempts reached - return No route
+ * }
+ * Upstream upstream = get(); // get current upstream and send request
+ * if (error) { // if request failed with 5xx or 429 - report current upstream as failed and call next for retry
+ *     failed();
+ *     next(); // retry
+ * } else {
+ *     succeed(); // if 200 - report upstream as succeeded
+ * }
+ * </pre>
+ */
+@Slf4j
+public class UpstreamRoute {
 
-    private final List<Upstream> upstreams;
-    private final int offset;
+    private static final long DEFAULT_RETRY_AFTER_SECONDS_VALUE = 30;
+
+    private final LoadBalancer<UpstreamState> balancer;
     /**
-     * The maximum number of retries for all upstreams.
+     * The maximum number of upstreams this route can use due to retries.
      */
-    private final int maxRetries;
-
-    private Upstream upstream;
-    private int retries;
-    private int next;
-    private int prev;
+    private final int maxUpstreamsToUse;
 
     /**
-     * @return the number of upstreams which returned any http response.
+     * Current upstream state
      */
-    public int used() {
-        return next;
+    @Nullable
+    private UpstreamState upstreamState;
+    private int used;
+
+    public UpstreamRoute(LoadBalancer<UpstreamState> balancer, int maxUpstreamsToUse) {
+        this.balancer = balancer;
+        this.maxUpstreamsToUse = maxUpstreamsToUse;
+        this.upstreamState = balancer.next();
     }
 
     /**
-     * @return the number of retries due to connection errors.
+     * @return the number of used upstreams.
      */
-    public int retries() {
-        return Math.min(retries, maxRetries);
+    public int used() {
+        return used + 1;
```

Comment:
I still don't get it why we return used + 1 instead of just the actual number of used?

Reply:
because we already call next in constructor

Reply:
The previous concept Iterator was more clear.
You could increase the var used in the constructor as well?

Reply:
previous iterator and balancer was simpler, we can initialize with 1 in constructor